features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.cpp

```diff
@@ -0,0 +1,168 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "os_wrapper/thread.h"
+#include "os_wrapper/mutex.h"
+#include "os_wrapper/semaphore.h"
+#include "rtos/Thread.h"
+#include "rtos/ThisThread.h"
+#include "rtos/Semaphore.h"
+#include "rtos/Mutex.h"
+
+extern "C" void *os_wrapper_semaphore_create(uint32_t max_count, uint32_t initial_count,
+                                             const char *name)
+{
+    if (initial_count > max_count) {
+        return NULL;
+    }
+
+    return (void *)new rtos::Semaphore(initial_count, max_count);
+}
+
+extern "C" uint32_t os_wrapper_semaphore_acquire(void *handle, uint32_t timeout)
+{
+    bool status;
+    rtos::Semaphore *new_sema = (rtos::Semaphore *)handle;
+
+    status = new_sema->try_acquire_for((timeout == OS_WRAPPER_WAIT_FOREVER) ?
+                                       osWaitForever : timeout);
+    if (false == status) {
+        return OS_WRAPPER_ERROR;
+    }
+
+    return OS_WRAPPER_SUCCESS;
+}
+
+extern "C" uint32_t os_wrapper_semaphore_release(void *handle)
+{
+    osStatus status;
+    rtos::Semaphore *new_sema = (rtos::Semaphore *)handle;
+
+    status = new_sema->release();
+    if (status != osOK) {
+        return OS_WRAPPER_ERROR;
+    }
+
+    return OS_WRAPPER_SUCCESS;
+}
+
+extern "C" uint32_t os_wrapper_semaphore_delete(void *handle)
+{
+    delete (rtos::Semaphore *)handle;
+
+    return OS_WRAPPER_SUCCESS;
+}
+
+extern "C" void *os_wrapper_mutex_create(void)
+{
+    return (void *)new rtos::Mutex();
+}
+
+extern "C" uint32_t os_wrapper_mutex_acquire(void *handle, uint32_t timeout)
+{
+    bool status;
+    rtos::Mutex *new_mutex = (rtos::Mutex *)handle;
+
+    status = new_mutex->trylock_for((timeout == OS_WRAPPER_WAIT_FOREVER) ?
+                                    osWaitForever : timeout);
+    if (!status) {
+        return OS_WRAPPER_ERROR;
+    }
+
+    return OS_WRAPPER_SUCCESS;
+}
+
+extern "C" uint32_t os_wrapper_mutex_release(void *handle)
+{
+    osStatus status = osOK;
+    rtos::Mutex *new_mutex = (rtos::Mutex *)handle;
+
+    status = new_mutex->unlock();
+    if (status != osOK) {
+        return OS_WRAPPER_ERROR;
+    }
+
+    return OS_WRAPPER_SUCCESS;
+}
+
+extern "C" uint32_t os_wrapper_mutex_delete(void *handle)
+{
+    delete (rtos::Mutex *)handle;
+
+    return OS_WRAPPER_SUCCESS;
+}
+
+extern "C" void *os_wrapper_thread_new(const char *name, int32_t stack_size,
+                                       os_wrapper_thread_func func, void *arg,
+                                       uint32_t priority)
+{
+    rtos::Thread *new_thread = NULL;
+    /* This enables thread context management for trustzone */
+    uint32_t tz_module = 1;
+
+    new_thread = new rtos::Thread(tz_module, (osPriority_t) priority, stack_size, NULL, name);
```

Comment:
TF-M implementation of the wrapper has check for stack_size == -1 for default stack size, therefore it might be possible that some of the tests are passing the stack_size as -1. Therefore it is necessary to check the stack_size before passing it to Thread constructor.

Reply:
I can check the `stack_size` to -1 but if it is then can't allocate anything as there is NO default size set, would be ok to return? At the moment if -1 is used then class doesn't have default size - https://github.com/ARMmbed/mbed-os/blob/feature_twincpu/rtos/source/Thread.cpp#L45

Reply:
We could use the Mbed OS C++ Thread class's default stack size, when -1 is passed by wrapper users, by not specifying the stack size to create in the constructor. This would provide an implementation of the wrappers that does what wrapper users expect.

Reply:
Oops. Seems I'm imagining constructors that don't exist. Anyhow, there ought to be an Mbed default stack size macro we can use.

Reply:
Possibly `OS_STACK_SIZE` as defined by `rtos/source/TARGET_CORTEX/mbed_rtx_conf.h`, which reads from the Mbed configuration system (i.e. all that json stuff).

Reply:
Actually yes, there is such a constructor. We should use it. 

Before, I was reading the cpp instead of the header file.

See https://github.com/ARMmbed/mbed-os/blob/feature_twincpu/rtos/Thread.h#L100-L105

Reply:
Yes, that can be done - please note - `OS_STACK_SIZE` is **4096 bytes**. Fixed.